disktable.go

```diff
@@ -0,0 +1,650 @@
+				panic("known key search came up with different key: HUGE PROBLEM, HUGE!!!!!!")
+			}
+			value, err := item.ValueCopy(nil)
+			if err != nil {
+				return fmt.Errorf("problem getting value of key(%s): %w", string(key), err)
+			}
+			ch <- Result{Value: value}
+		}
+		return nil
+	}
+}
+
+type lookupResult struct {
+	i   uint64
+	err error
+}
+
+// indexLookup returns all matching keys in an index.
+func (t *Table) indexLookup(indexName string, key []byte) chan lookupResult {
+	index := t.indexByName[indexName]
+
+	realKey := finalKey(key, 0)
+
+	ch := make(chan lookupResult, 1)
+	if index.AllowDuplicates {
+		// multiMatch will close the channel.
+		go t.multiMatch(realKey, index, ch)
+	} else {
+		ch <- t.exactMatch(realKey, index)
+		close(ch)
+	}
+
+	return ch
+}
+
+// exactMatch does an exact match for a key in the index. key must have been finalKey()'d.
+func (t *Table) exactMatch(key []byte, index *Index) lookupResult {
+	var result lookupResult
+
+	f := func(txn *badger.Txn) error {
+		item, err := txn.Get(key)
+		if err != nil {
+			return err
+		}
+		if !bytes.Equal(item.Key(), key) {
+			panic("known key search came up with different key: HUGE PROBLEM, HUGE!!!!!!")
+		}
+		value, err := item.ValueCopy(nil)
+		if err != nil {
+			return fmt.Errorf("problem getting value of key(%s): %w", string(key), err)
+		}
+		result = lookupResult{i: endian.Uint64(value)}
+		return nil
+	}
+
+	if err := index.db.View(f); err != nil {
+		result.err = err
+	}
+	return result
+
+}
+
+// multiMatch matches all keys in an index that allows duplicates. key must have been finalKey()'d.
+func (t *Table) multiMatch(key []byte, index *Index, ch chan lookupResult) {
+	defer close(ch)
+
+	keyData, err := getPrefixSuffix(key)
+	if err != nil {
+		ch <- lookupResult{err: err}
+		return
+	}
+
+	tempBuff := make([]byte, 8)
+	err = index.db.View(
+		func(txn *badger.Txn) error {
+			it := txn.NewIterator(badger.DefaultIteratorOptions)
+			defer it.Close()
+
+			for it.Seek(keyData.prefix); it.ValidForPrefix(keyData.prefix); it.Next() {
+				item := it.Item()
+
+				foundKeyData, err := getPrefixSuffix(item.Key())
+				if err != nil {
+					return fmt.Errorf("multiMatch of index(%s) had bad key(%v): %w", index.Name, item.Key(), err)
+				}
+				// It is possible to have a prefix match for something not exactly the same.
+				// But we are looking for exact prefix matches followed by a number followed by
+				// our key size. By matching prefix and key size, we can be sure of an exact match.
+				// If not, throw out the data.
+				if len(foundKeyData.prefix) != len(keyData.prefix) {
+					continue
+				}
+				b, err := item.ValueCopy(tempBuff)
+				if err != nil {
+					return err
+				}
+				ch <- lookupResult{i: endian.Uint64(b)}
+			}
+			return nil
+		},
+	)
+	if err != nil {
+		ch <- lookupResult{err: err}
+	}
+}
+
+func (t *Table) FetchAll(ctx context.Context) (chan Result, error) {
+	stream := t.primary.db.NewStream()
+	stream.NumGo = 16
```

Comment:
worth making this configurable?

Reply:
Honestly, can't say.  But I went ahead and added an optional argument to allow it to be adjusted.